### ISSUE TITLE
Controller detection

### ISSUE THREAD

**Issue by CpazR** [vitor251093/KHDays_FM#32]
Maybe a long term concern.

When building and running for the first time, everything is unassigned to the controller. 

Ideally, it should auto-detect any connected controller on either initial loading of the game, or whenever a controller is connected (assuming the mappings are supposed to be static).

**Comment by vitor251093** [comment]
I agree. Although, I'm not sure if the mappings values would be the same for every controller (probably not), which creates a bit of an issue. If we can find a way to inherit those mappings from a common place, or at least add automatic support for some of the most common controllers (like PS3, PS4, PS5, Xbox 360, Xbox One and Xbox Series X), it would probably be a good start.

**Comment by vitor251093** [comment]
I'm not sure if I understood the suggestion. Could you give an example?

**Comment by CpazR** [comment]
Again, don't know how MelonDS handles it's input, so maybe this is unrealistic, but hypothetically on startup, it would load custom textures for the button prompts used for the shortcut menu as an example:
Instead of X, Y, B (can't remember how the shortcuts are laid out lol) it would have triangle, square, and a cross if it detects a dualshock.

These textures in theory could be custom and packaged with the executable because of that.

**Comment by vitor251093** [comment]
Ohhhh, ok, now I get it.

It isn't unrealistic actually. However, based my experience with HID devices, I can tell you that controllers usually don't inform the name of their inputs. They are identified by indexes. Which means that, just like with the controller detection, we wouldn't be able to get the key names out of the box; that would either need to be configured manually from within Melon Mix, or automatically detected in case the controller is one of the common ones mentioned before.

**Comment by sandwichwater** [comment]
for the time being we can probably include folders with the different  button prompt textures along with the game , like xbox , ps , ds , etc , then have the user just drag and drop the texture files into their texture folder . probably the easiest  way to solve this until automatic detection is figured out

**Comment by vitor251093** [comment]
I've opened a new issue to deal with the texture replacement topic (#77) so this issue can focus on the controller automatic detection

**Comment by vitor251093** [comment]
I've made a research, and it seems like using the controller name may be the best approach.

https://discourse.libsdl.org/t/joystick-identification/18653

We can map controller names to a set of default mappings for each button/axis/hat, so known controllers work by default.

I thought about leaving that map on an external file, but it will thinking about it, I think it would be unnecessarily more complicated with almost no advantages over just leaving that map inside the code. Also, extracting that to an external file in the future would be possible, but going the other way around would be complicated.

**Comment by vitor251093** [comment]
While the default mappings should be in the code, the joystick mappings should be moved from the ini file to a separated file, so different controls connected to the same computer may have different mappings.

**Comment by vitor251093** [comment]
Development started at branch `base/multiple-joystick-mappings`

**Comment by vitor251093** [comment]
I decided to change the approach, and go straight to the controller mapping. I've added automatic mapping for Xbox 360 Wired controllers. I think it would be interesting to support automatically the controllers from:
- Xbox 360 (partially done)
- Xbox One
- Xbox Series S/X
- PS3
- PS4
- PS5
- SteamDeck

**Comment by CpazR** [comment]
> I decided to change the approach, and go straight to the controller mapping. I've added automatic mapping for Xbox 360 Wired controllers. I think it would be interesting to support automatically the controllers from:
> 
> * Xbox 360 (wireless; wired one is already supported)
> * Xbox One
> * Xbox Series S/X
> * PS3
> * PS4 (v1 and v2)
> * PS5
> * SteamDeck
> 
> I have a PS4 v2 controller, but can't test it right now.

Does steam input have dedicated mappings for its controller types? Would be handy to have something for more generic controllers.

**Comment by vitor251093** [comment]
> Does steam input have dedicated mappings for its controller types?

I'm not sure if I understood the question.

> Would be handy to have something for more generic controllers.

Those unfortunately vary a lot. I _can_ create a system that attaches them automatically based in the connected controller buttons, but they may end up wildly wrong.

One example: many D-Pads work as a single input which goes from 0 to 7, with each value representing one direction, including intermediate ones. Other D-Pads work as 4 buttons, of the same kind used by the buttons on the right side of controllers. So identifying D-pads among buttons is complicated on some controllers.

Because of this, the most simple approach would be to configure the most common controllers, and allowing people to manually configure generic ones.

**Comment by vitor251093** [comment]
> Does steam input have dedicated mappings for its controller types?

Oh, wait, you mean this?

https://partner.steamgames.com/doc/features/steam_controller/device

**Comment by CpazR** [comment]
> > Does steam input have dedicated mappings for its controller types?
> 
> Oh, wait, you mean this?
> 
> https://partner.steamgames.com/doc/features/steam_controller/device

I think so. Steam input is a wrapper anyway. Maybe it just binds things to xinput???

I can't remember off the top of my head.

**Comment by vitor251093** [comment]
Regarding DirectInput controller, I think it tries to map automatically, in an approach similar to what I mentioned I could do before, although, with Valve-budget xD

But regarding Xinput controllers, maybe they can be attributed automatically. Idk, I'm not used to those.

**Comment by vitor251093** [comment]
I got back to work on this one, and started from scratch.

The first necessary thing was to implement individual mapping for different controller models, which I made (still being tested) and opened a PR to melonDS with the feature.

https://github.com/melonDS-emu/melonDS/pull/2569

The next step is adding the automatic mappings of each controller to the plugins of each game, so they are applied when launching the game through the 1.5+2.5 collection (where manual mapping becomes less natural, and, therefore makes it the best place to start with this feature).

I already added the mapping for PS4 v2 and Steam input controllers, but it would be good to complete the original list, if possible, + Steam Input. Btw, I didn't know how Steam Input worked back in the day, and that's why I got confused when you talked about it xD

- Xbox 360 (wired)
- Xbox 360 (wireless)
- Xbox One
- Xbox Series S/X
- PS3
- PS4 v1
- PS4 v2 - **DONE**
- PS5
- SteamDeck
- Steam Input - **DONE**
